core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -700,19 +764,16 @@ class Partition(val topicPartition: TopicPartition,
       inWriteLock(leaderIsrUpdateLock) {
         // check if this replica needs to be added to the ISR
         if (needsExpandIsr(followerReplica)) {
-          val newInSyncReplicaIds = inSyncReplicaIds + followerReplica.brokerId
-          info(s"Expanding ISR from ${inSyncReplicaIds.mkString(",")} to ${newInSyncReplicaIds.mkString(",")}")
-          // update ISR in ZK and cache
-          expandIsr(newInSyncReplicaIds)
+          expandIsr(followerReplica.brokerId)
         }
       }
     }
   }
 
   private def needsExpandIsr(followerReplica: Replica): Boolean = {
-    leaderLogIfLocal.exists { leaderLog =>
+    !hasInFlightAlterIsr && leaderLogIfLocal.exists { leaderLog =>
```

Comment:
I think we can refactor this a little bit to avoid some duplication and inconsistency. We have the following logic above when updating follower state:
```scala
        if (!isrState.maximalIsr.contains(followerId))
          maybeExpandIsr(followerReplica, followerFetchTimeMs)
```
This is a little inconsistent because here we are checking `isrState.isr`. I'd suggest splitting this method into something like the following:

```scala
def hasReachedHighWatermark(follower: Replica): Boolean = {
  leaderLogIfLocal.exists { leaderLog =>
    val leaderHighwatermark = leaderLog.highWatermark
    isFollowerInSync(follower, leaderHighwatermark)
  }
}

def canAddToIsr(followerId: Int): Boolean = {
  val current = isrState
  !current.inflight && !current.isr.contains(followerId)
}

def needsExpandIsr(follower: Replica): Boolean = {
  canAddToIsr(follower.brokerId) && hasReachedHighWatermark(follower)
}
```

Then we can change the logic in `maybeExpandIsr` to the following:
```scala
    val needsIsrUpdate = canAddToIsr(followerReplica) && inReadLock(leaderIsrUpdateLock) {
...
```